In [ ]:
!pip install --upgrade -q "transformers[torch]" accelerate datasets evaluate decord

# 1. Import Dependensi

Dataset menggunakan dari MERL yang udah kuupload di huggingface

In [1]:
import os
import shutil
from huggingface_hub import snapshot_download

WORK_DIR = "/kaggle/working"
DATASET_DIR = os.path.join(WORK_DIR, "MERL_Shopping_Dataset")
VIDEOS_DIR = os.path.join(DATASET_DIR, "videos")
LABELS_DIR = os.path.join(DATASET_DIR, "labels")

# --- Download Dataset ---
if not os.path.exists(DATASET_DIR):
    print("Downloading dataset from HuggingFace...")
    snapshot_download(
        repo_id="haipradana/merl-shopping-action-detection",
        repo_type="dataset",
        local_dir=DATASET_DIR
    )
    print("Dataset downloaded!")
else:
    print("Dataset already exists in working directory.")

Fetching 214 files:   0%|          | 0/214 [00:00<?, ?it/s]

28_3_label.mat:   0%|          | 0.00/420 [00:00<?, ?B/s]

27_2_label.mat:   0%|          | 0.00/437 [00:00<?, ?B/s]

28_2_label.mat:   0%|          | 0.00/363 [00:00<?, ?B/s]

27_3_label.mat:   0%|          | 0.00/472 [00:00<?, ?B/s]

28_1_label.mat:   0%|          | 0.00/406 [00:00<?, ?B/s]

30_2_label.mat:   0%|          | 0.00/391 [00:00<?, ?B/s]

30_1_label.mat:   0%|          | 0.00/478 [00:00<?, ?B/s]

29_2_label.mat:   0%|          | 0.00/435 [00:00<?, ?B/s]

29_3_label.mat:   0%|          | 0.00/385 [00:00<?, ?B/s]

29_1_label.mat:   0%|          | 0.00/364 [00:00<?, ?B/s]

31_2_label.mat:   0%|          | 0.00/563 [00:00<?, ?B/s]

30_3_label.mat:   0%|          | 0.00/430 [00:00<?, ?B/s]

31_1_label.mat:   0%|          | 0.00/440 [00:00<?, ?B/s]

31_3_label.mat:   0%|          | 0.00/406 [00:00<?, ?B/s]

32_2_label.mat:   0%|          | 0.00/496 [00:00<?, ?B/s]

32_1_label.mat:   0%|          | 0.00/472 [00:00<?, ?B/s]

32_3_label.mat:   0%|          | 0.00/465 [00:00<?, ?B/s]

34_1_label.mat:   0%|          | 0.00/405 [00:00<?, ?B/s]

35_1_label.mat:   0%|          | 0.00/375 [00:00<?, ?B/s]

33_1_label.mat:   0%|          | 0.00/458 [00:00<?, ?B/s]

9_1_label.mat:   0%|          | 0.00/444 [00:00<?, ?B/s]

21_2_label.mat:   0%|          | 0.00/451 [00:00<?, ?B/s]

21_1_label.mat:   0%|          | 0.00/435 [00:00<?, ?B/s]

23_1_label.mat:   0%|          | 0.00/433 [00:00<?, ?B/s]

22_3_label.mat:   0%|          | 0.00/425 [00:00<?, ?B/s]

22_1_label.mat:   0%|          | 0.00/368 [00:00<?, ?B/s]

22_2_label.mat:   0%|          | 0.00/374 [00:00<?, ?B/s]

23_2_label.mat:   0%|          | 0.00/464 [00:00<?, ?B/s]

21_3_label.mat:   0%|          | 0.00/425 [00:00<?, ?B/s]

1_1_crop.mp4:   0%|          | 0.00/15.7M [00:00<?, ?B/s]

1_2_crop.mp4:   0%|          | 0.00/15.3M [00:00<?, ?B/s]

1_3_crop.mp4:   0%|          | 0.00/22.7M [00:00<?, ?B/s]

20_1_crop.mp4:   0%|          | 0.00/16.0M [00:00<?, ?B/s]

20_2_crop.mp4:   0%|          | 0.00/21.7M [00:00<?, ?B/s]

2_1_crop.mp4:   0%|          | 0.00/15.6M [00:00<?, ?B/s]

2_2_crop.mp4:   0%|          | 0.00/16.5M [00:00<?, ?B/s]

20_3_crop.mp4:   0%|          | 0.00/20.3M [00:00<?, ?B/s]

2_3_crop.mp4:   0%|          | 0.00/16.0M [00:00<?, ?B/s]

3_1_crop.mp4:   0%|          | 0.00/13.9M [00:00<?, ?B/s]

3_2_crop.mp4:   0%|          | 0.00/19.5M [00:00<?, ?B/s]

3_3_crop.mp4:   0%|          | 0.00/15.3M [00:00<?, ?B/s]

4_1_crop.mp4:   0%|          | 0.00/15.2M [00:00<?, ?B/s]

4_3_crop.mp4:   0%|          | 0.00/16.4M [00:00<?, ?B/s]

4_2_crop.mp4:   0%|          | 0.00/15.7M [00:00<?, ?B/s]

5_1_crop.mp4:   0%|          | 0.00/15.2M [00:00<?, ?B/s]

5_2_crop.mp4:   0%|          | 0.00/17.7M [00:00<?, ?B/s]

5_3_crop.mp4:   0%|          | 0.00/18.8M [00:00<?, ?B/s]

6_1_crop.mp4:   0%|          | 0.00/15.6M [00:00<?, ?B/s]

6_3_crop.mp4:   0%|          | 0.00/17.8M [00:00<?, ?B/s]

6_2_crop.mp4:   0%|          | 0.00/16.6M [00:00<?, ?B/s]

7_1_crop.mp4:   0%|          | 0.00/14.8M [00:00<?, ?B/s]

7_2_crop.mp4:   0%|          | 0.00/16.7M [00:00<?, ?B/s]

7_3_crop.mp4:   0%|          | 0.00/17.5M [00:00<?, ?B/s]

8_2_crop.mp4:   0%|          | 0.00/19.4M [00:00<?, ?B/s]

8_1_crop.mp4:   0%|          | 0.00/14.1M [00:00<?, ?B/s]

8_3_crop.mp4:   0%|          | 0.00/17.7M [00:00<?, ?B/s]

9_1_crop.mp4:   0%|          | 0.00/15.3M [00:00<?, ?B/s]

9_2_crop.mp4:   0%|          | 0.00/16.2M [00:00<?, ?B/s]

9_3_crop.mp4:   0%|          | 0.00/17.2M [00:00<?, ?B/s]

21_1_crop.mp4:   0%|          | 0.00/15.8M [00:00<?, ?B/s]

21_3_crop.mp4:   0%|          | 0.00/18.7M [00:00<?, ?B/s]

22_1_crop.mp4:   0%|          | 0.00/18.1M [00:00<?, ?B/s]

21_2_crop.mp4:   0%|          | 0.00/17.0M [00:00<?, ?B/s]

22_2_crop.mp4:   0%|          | 0.00/17.5M [00:00<?, ?B/s]

22_3_crop.mp4:   0%|          | 0.00/18.0M [00:00<?, ?B/s]

23_1_crop.mp4:   0%|          | 0.00/16.9M [00:00<?, ?B/s]

23_2_crop.mp4:   0%|          | 0.00/21.9M [00:00<?, ?B/s]

23_3_crop.mp4:   0%|          | 0.00/26.7M [00:00<?, ?B/s]

24_1_crop.mp4:   0%|          | 0.00/17.9M [00:00<?, ?B/s]

24_2_crop.mp4:   0%|          | 0.00/19.7M [00:00<?, ?B/s]

24_3_crop.mp4:   0%|          | 0.00/21.4M [00:00<?, ?B/s]

25_1_crop.mp4:   0%|          | 0.00/17.3M [00:00<?, ?B/s]

25_2_crop.mp4:   0%|          | 0.00/19.1M [00:00<?, ?B/s]

25_3_crop.mp4:   0%|          | 0.00/17.6M [00:00<?, ?B/s]

26_1_crop.mp4:   0%|          | 0.00/17.4M [00:00<?, ?B/s]

26_2_crop.mp4:   0%|          | 0.00/18.8M [00:00<?, ?B/s]

26_3_crop.mp4:   0%|          | 0.00/16.4M [00:00<?, ?B/s]

Dataset downloaded!


# 2. Preprocessing

## Preprocessing Dataset

datasete dimerge antara train/test/split supaya memudahkan dalam memuat klip video dari frame dengan decord dengan fungsi create_clips_dataframe. Nanti displit lagi.

In [2]:
# --- Gabungkan Folder Train/Val/Test (jika ada) ---
if os.path.exists(os.path.join(VIDEOS_DIR, "train")):
    print("Pre-split folders found. Merging them into single directories...")
    # Buat direktori sementara untuk penggabungan
    merged_videos_dir = f"{DATASET_DIR}/videos_merged"
    merged_labels_dir = f"{DATASET_DIR}/labels_merged"
    os.makedirs(merged_videos_dir, exist_ok=True)
    os.makedirs(merged_labels_dir, exist_ok=True)

    for split in ['train', 'val', 'test']:
        # Gabungkan video
        split_videos_path = os.path.join(VIDEOS_DIR, split)
        if os.path.exists(split_videos_path):
            for f in os.listdir(split_videos_path):
                shutil.copy(os.path.join(split_videos_path, f), merged_videos_dir)
        # Gabungkan label
        split_labels_path = os.path.join(LABELS_DIR, split)
        if os.path.exists(split_labels_path):
            for f in os.listdir(split_labels_path):
                shutil.copy(os.path.join(split_labels_path, f), merged_labels_dir)

    # Ganti folder lama dengan yang sudah digabung
    shutil.rmtree(VIDEOS_DIR)
    shutil.rmtree(LABELS_DIR)
    shutil.move(merged_videos_dir, VIDEOS_DIR)
    shutil.move(merged_labels_dir, LABELS_DIR)
    print("Merging complete.")

Pre-split folders found. Merging them into single directories...
Merging complete.


## Preprocessing label dan Membuat Clips

1. Load datset label .mat dengan scipy.io loadmat. Ini membaca file label untuk mengetahui di frame berapa saja sebuah aksi terjadi (misalnya, "Reach To Shelf" terjadi dari frame 100 hingga 150).
2. Membuat fungsi create_clips_dataframe(videos_dir, labels_dir), untuk memetakan video dan labelnya. Lalu membuat sebuah daftar (yang kemudian menjadi Pandas DataFrame) yang berisi instruksi. Setiap baris dalam DataFrame itu adalah satu "klip virtual" yang didefinisikan oleh:
- video_id: Video mana yang harus digunakan.
- start_frame: Mulai dari frame ke berapa.
- end_frame: Berhenti di frame ke berapa.
- class: Apa nama aksi untuk segmen frame tersebut.

((lihat output))

In [ ]:
import pandas as pd
import glob
import re
from scipy.io import loadmat # fungsi khusus dari matlab untuk membaca format mat
import numpy as np

# Fungsi untuk load label .mat
def load_mat_labels(label_path):
    try:
        mat_data = loadmat(label_path) # mengubah ke dict
        if 'tlabs' in mat_data: return mat_data['tlabs']
        data_keys = [k for k in mat_data.keys() if not k.startswith('__')]
        if data_keys: return mat_data[data_keys[0]]
        return None
    except Exception as e:
        print(f"Error loading {label_path}: {e}")
        return None

# Fungsi baru untuk membuat DataFrame berisi informasi klip
def create_clips_dataframe(videos_dir, labels_dir):
    all_clips_data = []
    video_paths = sorted(glob.glob(os.path.join(videos_dir, "*_crop.mp4")))
    
    # Kita butuh FPS untuk konversi frame nanti. Asumsikan 30 FPS.
    # Jika berbeda, decord akan membacanya nanti.
    fps_skip = 2 # Dari 30fps di video asli ke 15fps target

    for video_path in video_paths:
        match = re.search(r'(\d+)_(\d+)_crop\.mp4', os.path.basename(video_path))
        if not match: continue
        
        base_id, sub_id = match.groups()
        video_id_str = f"{base_id}_{sub_id}"
        
        label_path = os.path.join(labels_dir, f"{video_id_str}_label.mat")
        if not os.path.exists(label_path): continue

        labels_data = load_mat_labels(label_path)
        if labels_data is None: continue

        # membuat interval aksi dari .mat
        action_intervals = []
        for class_idx, class_actions in enumerate(labels_data):
            if len(class_actions[0]) > 0:
                for start_frame, end_frame in class_actions[0]:
                    action_intervals.append({
                        'start': int(start_frame) - 1, # 0-based
                        'end': int(end_frame) - 1,   # 0-based
                        'class': class_idx + 1  # 1-5
                    })
        action_intervals.sort(key=lambda x: x['start'])

        # membuat klip, termasuk kelas BACKGROUND
        current_frame = 0
        # video_total_frames = 10000 # Placeholder, akan dibaca decord nanti
        
        for interval in action_intervals:
            if current_frame < interval['start']: # jika ada jeda (tidak ada aksi)
                all_clips_data.append({
                    'video_id': video_id_str,
                    'start_frame': current_frame,
                    'end_frame': interval['start'] - 1,
                    'class': 0 # Background
                })
            
            all_clips_data.append({
                'video_id': video_id_str,
                'start_frame': interval['start'],
                'end_frame': interval['end'],
                'class': interval['class']
            })
            current_frame = interval['end'] + 1
        
        # Sisa klip background di akhir
        # Kita tidak perlu ini sekarang, biarkan Dataset yang mengaturnya nanti
    
    return pd.DataFrame(all_clips_data)

# Buat DataFrame utama
print("Creating master DataFrame from labels...")
combined_df = create_clips_dataframe(VIDEOS_DIR, LABELS_DIR)

# Hapus klip yang terlalu pendek
min_clip_length = 16 # Minimal 16 frame @30fps, atau ~0.5 detik
combined_df['duration'] = combined_df['end_frame'] - combined_df['start_frame']
combined_df = combined_df[combined_df['duration'] >= min_clip_length].reset_index(drop=True)

print(f"DataFrame created with {len(combined_df)} clips.")
print("\nClass Distribution:")
print(combined_df['class'].value_counts().sort_index())
combined_df.head()

Creating master DataFrame from labels...
DataFrame created with 5648 clips.

Class Distribution:
class
0     460
1    1669
2    1559
3     506
4     655
5     799
Name: count, dtype: int64


video_id  start_frame  end_frame  class  duration
0     10_1            0        176      0       176
1     10_1          177        229      5        52
2     10_1          235        256      1        21
3     10_1          256        290      2        34
4     10_1          299        525      4       226

## Split dataset

In [4]:
from sklearn.model_selection import train_test_split

# Split data 70% train, 15% validation, 15% test
train_df, temp_df = train_test_split(
    combined_df,
    test_size=0.3,
    stratify=combined_df['class'],
    random_state=42
)

val_df, test_df = train_test_split(
    temp_df,
    test_size=0.5,
    stratify=temp_df['class'],
    random_state=42
)

print(f"Data split successfully:")
print(f"  Training set:   {len(train_df)} clips")
print(f"  Validation set: {len(val_df)} clips")
print(f"  Test set:       {len(test_df)} clips")

# train_df_subset = train_df.sample(frac=0.5, random_state=42)
# val_df_subset = val_df.sample(frac=0.5, random_state=42)

Data split successfully:
  Training set:   3953 clips
  Validation set: 847 clips
  Test set:       848 clips


# 3. Utilitas Pembacaan Segmen Video

fungsi MerlVideoDataset() ini cukup penting, bisa dibilang dia yg "memotong", sebenrnya cuma membaca segmen video dengan decord (lebih fleksibel dan efisien).

di kode cukup jelas

In [ ]:
import torch
from torch.utils.data import Dataset
from decord import VideoReader, cpu

class MerlVideoDataset(Dataset):
    def __init__(self, df, image_processor, videos_dir, num_frames=16):
        self.df = df
        self.image_processor = image_processor
        self.videos_dir = videos_dir
        self.num_frames = num_frames

    def __len__(self):
        return len(self.df)

    def __getitem__(self, idx):
        row = self.df.iloc[idx]
        video_path = os.path.join(self.videos_dir, f"{row['video_id']}_crop.mp4")

        # Baca video dengan decord
        vr = VideoReader(video_path, ctx=cpu(0))
        total_frames_in_video = len(vr)

        # Ambil frame indices dari klip yang didefinisikan
        start_frame = row['start_frame'] # <- Membaca instruksi frame awal
        end_frame = row['end_frame'] # <- Membaca instruksi frame akhir
        
        # Buat sampel frame secara merata di dalam klip
        indices = np.linspace(start_frame, end_frame, num=self.num_frames, dtype=int)
        indices = np.clip(indices, 0, total_frames_in_video - 1)

        # Ambil frame yang dibutuhkan secara efisien
        frames = vr.get_batch(indices).asnumpy() # <- di sini frame dibaca

        # Proses menggunakan processor dari Hugging Face (normalisasi, resize)
        processed_frames = self.image_processor(list(frames), return_tensors="pt")

        return {
            "pixel_values": processed_frames.pixel_values.squeeze(0), # Hapus dimensi batch
            "labels": torch.tensor(row['class'], dtype=torch.long)
        }

# 4. Hyperparameter for TrainingArgs

1. Menggunakan class weight untuk berusaha menyeimbangkan imbalanced data
2. Mendefinisikan CustomTrainer yang inheritance dari Trainer, sedikit ditambahkan supaya loss function dari CrossEntropyLoss bisa menggunakan bobot dari weight yg sebelumnya dibahas.
3. Mendefinisikan paramater untuk training
4. Inisiasi trainer dengan Custom Trainer

In [ ]:
import torch
from torch import nn
from transformers import AutoImageProcessor, AutoModelForVideoClassification, TrainingArguments, Trainer
import evaluate
import numpy as np

# Definisikan Bobot Kelas (Class Weights)
print("Menghitung bobot kelas untuk mengatasi data tidak seimbang...")
class_counts = combined_df['class'].value_counts().sort_index().values
total_samples = class_counts.sum()
num_classes = len(class_counts)
class_weights = total_samples / (num_classes * class_counts)
# Ubah ke tensor PyTorch dan pindahkan ke GPU
class_weights_tensor = torch.tensor(class_weights, dtype=torch.float).to("cuda")

print(f"Bobot Kelas yang akan digunakan: {class_weights_tensor.cpu().numpy()}")


# Buat Custom Trainer untuk Menggunakan Bobot 
class CustomTrainer(Trainer):
    def compute_loss(self, model, inputs, return_outputs=False, **kwargs):
        labels = inputs.pop("labels")
        # Forward pass
        outputs = model(**inputs)
        logits = outputs.get('logits')
        # Hitung loss dengan bobot yang sudah kita definisikan
        loss_fct = nn.CrossEntropyLoss(weight=class_weights_tensor)
        loss = loss_fct(logits.view(-1, self.model.config.num_labels), labels.view(-1))
        return (loss, outputs) if return_outputs else loss


# Definisikan label dan checkpoint model
class_names = ["Background", "Reach To Shelf", "Retract From Shelf", "Hand In Shelf", "Inspect Product", "Inspect Shelf"]
label2id = {name: i for i, name in enumerate(class_names)}
id2label = {i: name for i, name in enumerate(class_names)}
model_ckpt = "facebook/timesformer-base-finetuned-k400"

# Muat Image Processor dan Model
print("\n Memuat model TimeSformer dan image processor...")
image_processor = AutoImageProcessor.from_pretrained(model_ckpt)
model = AutoModelForVideoClassification.from_pretrained(
    model_ckpt,
    ignore_mismatched_sizes=True,
    num_labels=len(class_names),
    label2id=label2id,
    id2label=id2label
)

# Buat instance Dataset (gunakan kode dari langkah 5 Anda)
print(" Membuat instance PyTorch Dataset...")
train_dataset = MerlVideoDataset(train_df, image_processor, VIDEOS_DIR)
val_dataset = MerlVideoDataset(val_df, image_processor, VIDEOS_DIR)

# pake subset 50 persen dulu
# train_dataset = MerlVideoDataset(train_df_subset, image_processor, VIDEOS_DIR)
# val_dataset = MerlVideoDataset(val_df_subset, image_processor, VIDEOS_DIR)

# Definisikan Argumen Training
print(" Mengkonfigurasi argumen training...")
training_args = TrainingArguments(
    output_dir=os.path.join(WORK_DIR, "timesformer-merl-weighted-loss"),
    per_device_train_batch_size=2,  
    per_device_eval_batch_size=2,
    gradient_accumulation_steps=4,
    num_train_epochs=5, 
    learning_rate=3e-5, # Sedikit turunkan learning rate, kadang lebih stabil
    warmup_ratio=0.1,
    weight_decay=0.05,
    logging_strategy="epoch",
    eval_strategy="epoch",
    save_strategy="epoch",
    load_best_model_at_end=True,
    save_total_limit=2,
    fp16=True,
    report_to="none",
)

# Definisikan Fungsi Metrik
accuracy_metric = evaluate.load("accuracy")
def compute_metrics(eval_pred):
    predictions = np.argmax(eval_pred.predictions, axis=1)
    return accuracy_metric.compute(predictions=predictions, references=eval_pred.label_ids)

# Gunakan CustomTrainer, bukan Trainer biasa
print("Menginisialisasi CustomTrainer dengan class weights...")
trainer = CustomTrainer( # <-- Gunakan CustomTrainer
    model=model,
    args=training_args,
    train_dataset=train_dataset,
    eval_dataset=val_dataset,
    compute_metrics=compute_metrics,
    processing_class=image_processor,
)

# kenapa pake custom trainer? apakah karena karena ingi memanfaatkan classweight? YA

2025-06-30 14:06:54.785078: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1751292415.000882      92 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1751292415.063865      92 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered


⚖️ Menghitung bobot kelas untuk mengatasi data tidak seimbang...
Bobot Kelas yang akan digunakan: [2.0463767  0.5640104  0.60380584 1.8603425  1.4371501  1.1781393 ]

🔄 Memuat model TimeSformer dan image processor...


preprocessor_config.json:   0%|          | 0.00/412 [00:00<?, ?B/s]

Using a slow image processor as `use_fast` is unset and a slow processor was saved with this model. `use_fast=True` will be the default behavior in v4.52, even if the model was saved with a slow processor. This will result in minor differences in outputs. You'll still be able to use a slow processor with `use_fast=False`.


config.json: 0.00B [00:00, ?B/s]

pytorch_model.bin:   0%|          | 0.00/486M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/486M [00:00<?, ?B/s]

Some weights of TimesformerForVideoClassification were not initialized from the model checkpoint at facebook/timesformer-base-finetuned-k400 and are newly initialized because the shapes did not match:
- classifier.weight: found shape torch.Size([400, 768]) in the checkpoint and torch.Size([6, 768]) in the model instantiated
- classifier.bias: found shape torch.Size([400]) in the checkpoint and torch.Size([6]) in the model instantiated
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


📚 Membuat instance PyTorch Dataset...
⚙️ Mengkonfigurasi argumen training...


✨ Menginisialisasi CustomTrainer dengan class weights...


Total Sampel: 5648

Jumlah Kelas: 6

Bobot Kelas 0 = 5648 / (6 * 460) = 2.04

Bobot Kelas 1 = 5648 / (6 * 1669) = 0.56

Bobot Kelas 2 = 5648 / (6 * 1559) = 0.60

Bobot Kelas 3 = 5648 / (6 * 506) = 1.86

Bobot Kelas 4 = 5648 / (6 * 655) = 1.44

Bobot Kelas 5 = 5648 / (6 * 799) = 1.18

# 5. TRAINING

In [7]:
trainer.train()
print("training completed")

training completed


In [ ]:
!mkdir /kaggle/working/output

In [ ]:
import pandas as pd 

logs = trainer.state.log_history
log_df = pd.DataFrame(logs)
# %cd /kaggle/working
# !mkdir /outputnya/
output_dir = "/kaggle/working/output"
log_df.to_csv(os.path.join(output_dir, "training_logs.csv"), index=False)
print("Logs saved!")

# Test Inference Data

Aku udah upload dataset ke huggingface, jadi tinggal panggil.
2. Buat test_dataset
3. Predict, lalu hitung performanya

In [ ]:
from sklearn.metrics import classification_report, accuracy_score

model_trained_before = "haipradana/test-m-e-r-l"
device = "cuda" if torch.cuda.is_available() else "cpu"

test_image_processor = AutoImageProcessor.from_pretrained(model_trained_before)
test_model = AutoModelForVideoClassification.from_pretrained(model_trained_before).to(device)

test_dataset = MerlVideoDataset(
    df=test_df,
    image_processor=test_image_processor,
    videos_dir=VIDEOS_DIR
)

test_args = TrainingArguments(
    output_dir="./test_results",
    per_device_eval_batch_size=4, 
    report_to="none",
)

tester = Trainer(
    model=test_model,
    args=test_args,
)

print("\nMenjalankan prediksi pada test set...")
predictions = tester.predict(test_dataset)
print("Prediksi selesai")

# Ambil prediksi (logits) dan label sebenarnya
predicted_logits = predictions.predictions
true_labels = predictions.label_ids

# Ubah logits menjadi ID kelas yang diprediksi
predicted_labels = np.argmax(predicted_logits, axis=1)

# Dapatkan nama kelas dari `id2label` yang sudah didefinisikan sebelumnya
target_names = [id2label[i] for i in range(len(id2label))]

# Hitung dan cetak laporan klasifikasi (precision, recall, f1-score)
print("\n--- Laporan Klasifikasi ---")
print(classification_report(true_labels, predicted_labels, target_names=target_names))

# Hitung dan cetak akurasi keseluruhan
accuracy = accuracy_score(true_labels, predicted_labels)
print(f"Akurasi Keseluruhan: {accuracy:.4f}")

In [ ]:
# --- Laporan Klasifikasi TEST  ---
#                    precision    recall  f1-score   support

#        Background       0.54      0.68      0.60        69
#    Reach To Shelf       0.52      0.54      0.53       251
#Retract From Shelf       0.52      0.44      0.47       234
#     Hand In Shelf       0.61      0.71      0.65        76
#   Inspect Product       0.67      0.73      0.70        98
#     Inspect Shelf       0.84      0.75      0.79       120

#          accuracy                           0.59       848
#         macro avg       0.62      0.64      0.63       848
#      weighted avg       0.59      0.59      0.59       848

#Akurasi Keseluruhan: 0.5908

COnfusion matriks. model masih agak bingung dalam membedakan mana yang retract mana yang reach.

In [ ]:
from sklearn.metrics import confusion_matrix
import matplotlib.pyplot as plt
import seaborn as sns

cm = confusion_matrix(true_labels, predicted_labels, normalize='true')

plt.figure(figsize=(10, 8))
sns.heatmap(cm, annot=True, fmt='.2f', cmap='Blues', 
            xticklabels=class_names, yticklabels=class_names)
plt.title('Confusion Matrix pada Test Set')
plt.ylabel('Label Asli (True Label)')
plt.xlabel('Label Prediksi (Predicted Label)')
plt.show()

![Confusion Matrix](confusion_matrix_merl.png)